### PR TITLE
Reset gen dev

### PR DESCRIPTION
**Pull request by Yuan-Mao**
Current version of Zynq-Parrot can only be reset as a whole when the bistream is loaded. The only exception there is we have the ability to reset BP independently from the remaining of the entire system. In order to add a more general reset generation facility to Zynq-Parrot, this PR adds a bsg_tag_bitbang, a tag master and some tag clients under top_zynq so that PS can bitbang any reset sequence into the tag master through a CSR in zynq-shell.

Note that since we need zynq-shell to be reset before PS starts bit banging, the reset for zynq-shell (which is the reset for s00) is connected directly to the Xilinx reset generation IP, and the other resets (resets for s01, m00, m01) are from tag clients. Therefore after s00 has been reset, PS has to manually trigger the reset for s01, m00, m01 through the bitbang CSR. This can be seen in ps.cpp in this PR.

This PR also includes commits (ef0ff43 to d5619a5) from other PRs, which are https://github.com/black-parrot-hdk/zynq-parrot/pull/47, https://github.com/black-parrot-hdk/zynq-parrot/pull/49 and https://github.com/bespoke-silicon-group/basejump_stl/pull/618. Those commits should be removed when we want to merge this PR.

### CODE REVIEW

File: cosim/v/bsg_tag_bitbang.v [hunk 1]
```diff
@@ -0,0 +1,74 @@
+
```
Comment (dpetrisko) [inline cosim/v/bsg_tag_bitbang.v:1]: NOTE: use basejump version when merged

File: cosim/black-parrot-example/ps.cpp [hunk 2]
```diff
@@ -135,45 +183,67 @@ extern "C" void cosim_main(char *argstr) {
...
+              zpl->axil_read(GP0_RD_CSR_BITBANG       + gp0_addr_base),
+              zpl->axil_read(GP0_RD_CSR_DRAM_INITED + gp0_addr_base),
+              val = zpl->axil_read(GP0_RD_CSR_DRAM_BASE + gp0_addr_base));
+
+  // Reset bsg tag master
+  // This is needed because otherwise zeros_ctr_r in the tag master will never
+  //   be zero, though in hardware it should eventually be set to 0.
+
+  zpl->axil_write(GP0_WR_CSR_BITBANG + gp0_addr_base, 0x1, 0xf);
+  for(unsigned i = 0;i < get_bsg_tag_reset_len(NUM_RESET, 1);i++)
```
Comment (dpetrisko) [inline cosim/black-parrot-example/ps.cpp:196]: I wouldn't be this precise, just set to a really high number, like 1000

File: cosim/black-parrot-example/ps.cpp [hunk 3]
```diff
@@ -135,45 +183,67 @@ extern "C" void cosim_main(char *argstr) {
...
+    zpl->axil_write(GP0_WR_CSR_BITBANG + gp0_addr_base, 0x0, 0xf);
+  // Reset bsg client0
+  write_bsg_tag_packet(zpl, NUM_RESET, 1, 0, 0, -1U);
```
Comment (dpetrisko) [inline cosim/black-parrot-example/ps.cpp:199]: is -1U not undefined behavior?

File: cosim/black-parrot-example/v/bsg_tag_reset_gen.v [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+`include "bsg_defines.v"
+`include "bsg_tag.vh"
+
+/*
+ *   This is a simple wrapper that joins 1 bsg_tag_bitbang, 1 bsg_tag_master
+ * and els_p bsg_tag_clients. One bsg_tag_client for one reset generation.
+ *
+ */
+
+module bsg_tag_reset_gen
```
Comment (dpetrisko) [inline cosim/black-parrot-example/v/bsg_tag_reset_gen.v:11]: I don't think we gain much from this wrapper; let's just put in top_zynq

File: cosim/black-parrot-example/v/bsg_tag_reset_gen.v [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    ,.reset_i(tag_reset_i)
+    ,.data_i(tag_data_i)
+    ,.v_i(tag_v_i)
+    ,.ready_and_o() // UNUSED
+
+    ,.tag_clk_r_o(tag_clk_r_lo)
+    ,.tag_data_r_o(tag_data_r_lo)
+  );
+
+  bsg_tag_master #(
```
Comment (dpetrisko) [inline cosim/black-parrot-example/v/bsg_tag_reset_gen.v:46]: Use tag_master_decentralized, we're deprecating bsg_tag_master

File: cosim/black-parrot-example/v/bsg_tag_reset_gen.v [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+  `declare_bsg_tag_header_s(els_p, lg_width_lp)
+
+  bsg_tag_s [els_p-1:0] tag;
+  bsg_tag_bitbang bitbang (
+     .clk_i(tag_clk_i)
+    ,.reset_i(tag_reset_i)
+    ,.data_i(tag_data_i)
+    ,.v_i(tag_v_i)
+    ,.ready_and_o() // UNUSED
```
Comment (dpetrisko) [inline cosim/black-parrot-example/v/bsg_tag_reset_gen.v:40]: Can we add an assert that the packet is dropped if v_i & ~ready? I don't think it's possible to happen in practice, but still

File: cosim/black-parrot-example/ps.cpp [hunk 7]
```diff
@@ -93,6 +120,39 @@ inline uint64_t get_counter_64(bp_zynq_pl *zpl, uint64_t addr) {
...
+}
+#define safe_clog2(x) (((x) == 1U) ? 1U : clog2(x))
+unsigned get_bsg_tag_reset_len(unsigned els, unsigned width) {
+  unsigned lg_width = clog2(width + 1);
+  unsigned header_len = safe_clog2(els) + 1 + lg_width;
+  unsigned max_packet_len = ((1 << lg_width) - 1 + header_len);
+  return (1 << (safe_clog2(max_packet_len + 1))) + 1;
+}
+
+void write_bsg_tag_packet(bp_zynq_pl *zpl, unsigned els, unsigned width,
```
Comment (dpetrisko) [inline cosim/black-parrot-example/ps.cpp:139]: can we move this to include/common/bsg_tag_bitbang.h like we have for bsg_axil.h? That way it will act as a driver we can include in other examples

File: cosim/black-parrot-example/ps.cpp [hunk 8]
```diff
@@ -135,45 +183,67 @@ extern "C" void cosim_main(char *argstr) {
...
+  for(unsigned i = 0;i < get_bsg_tag_reset_len(NUM_RESET, 1);i++)
+    zpl->axil_write(GP0_WR_CSR_BITBANG + gp0_addr_base, 0x0, 0xf);
+  // Reset bsg client0
+  write_bsg_tag_packet(zpl, NUM_RESET, 1, 0, 0, -1U);
+  // Set bsg client0 to 1 (assert reset for axi buses m00, m01, s01)
+  write_bsg_tag_packet(zpl, NUM_RESET, 1, 1, 0, 0x1);
+  // Set bsg client0 to 0 (deassert reset for axi buses m00, m01, s01)
+  write_bsg_tag_packet(zpl, NUM_RESET, 1, 1, 0, 0x0);
+
+  // We need at least 4 additional toggles for data to propagate through
```
Comment (dpetrisko) [inline cosim/black-parrot-example/ps.cpp:205]: for the synchronizers or what?

File: cosim/black-parrot-example/ps.cpp [hunk 9]
```diff
@@ -135,45 +183,67 @@ extern "C" void cosim_main(char *argstr) {
...
+  for(int i = 0;i < 4;i++)
+    zpl->axil_write(GP0_WR_CSR_BITBANG + gp0_addr_base, 0x0, 0xf);
+
+#ifdef GP1_ENABLE
+  zpl->axi_gp1->reset(bp_zynq_pl::tick);
```
Comment (dpetrisko) [inline cosim/black-parrot-example/ps.cpp:210]: oh why are these necessary?

File: cosim/black-parrot-example/v/top.v [hunk 10]
```diff
@@ -487,17 +505,7 @@ module top
       ,.m01_axi_rready (m01_axi_rready)
       );
 
-`ifdef VERILATOR
```
Comment (dpetrisko) [inline cosim/black-parrot-example/v/top.v:490]: revert for tracing in verilator?

File: cosim/black-parrot-example/v/top.v [hunk 11]
```diff
@@ -514,6 +522,7 @@ module top
            $value$plusargs("c_args=%s", c_args);
          end
        cosim_main(c_args);
+       $finish;
```
Comment (dpetrisko) [inline cosim/black-parrot-example/v/top.v:525]: makes sense, but what issue did this fix?

File: cosim/black-parrot-example/v/top_zynq.sv [hunk 12]
```diff
@@ -188,10 +190,29 @@ module top_zynq
...
+   wire tag_v_li = csr_data_v_lo[0];
+   bsg_tag_reset_gen #(
+      .els_p(num_reset_lp)
+   ) reset_gen (
+      .tag_clk_i(aclk)
+     ,.tag_reset_i(~aresetn)
+     ,.tag_data_i(tag_data_li)
+     ,.tag_v_i(tag_v_li)
+     ,.clks_i({aclk})
+     ,.resets_o({resets_lo}) // assuming the PS programmer uses high-true signals
```
Comment (dpetrisko) [inline cosim/black-parrot-example/v/top_zynq.sv:207]: Let's keep the polarity the same as the PS, I think that's the least surprising implementation

File: cosim/black-parrot-example/v/top_zynq.sv [hunk 13]
```diff
@@ -188,10 +190,29 @@ module top_zynq
...
+   );
+
+   // When aresetn is asserted, s00 will be the only axi bus that has been initialized
+   assign s00_axi_aresetn = aresetn;
+   // Other axi buses need to be initialized through bit banging from PS
+   assign {s01_axi_aresetn, m00_axi_aresetn, m01_axi_aresetn} = {3{~resets_lo}};
```
Comment (dpetrisko) [inline cosim/black-parrot-example/v/top_zynq.sv:213]: Oh, I'm not sure we want to reset the AXI buses from this. I think that should stay as aresetn. It's very prone to deadlock if you mess with axi resets during runtime (basically if there's any outstanding transactions, the system will hang). I think only peripherals and be should be controlled by tag

File: cosim/include/vcs/bp_zynq_pl.h [hunk 14]
```diff
@@ -52,14 +52,13 @@ class zynq_scratchpad : public axil_device {
 };
 
 class bp_zynq_pl {
-
-  static std::unique_ptr<axilm<GP0_ADDR_WIDTH, GP0_DATA_WIDTH> > axi_gp0;
```
Comment (dpetrisko) [inline cosim/include/vcs/bp_zynq_pl.h:56]: why?

File: cosim/include/verilator/bp_zynq_pl.h [hunk 15]
```diff
@@ -48,27 +50,33 @@ class zynq_scratchpad : public axil_device {
 
 class bp_zynq_pl {
   static Vtop *tb;
+  static VerilatedFstC *wf;
```
Comment (dpetrisko) [inline cosim/include/verilator/bp_zynq_pl.h:53]: where is this initialized?

File: cosim/black-parrot-example/v/top.v (hunk 10 above)
Comment (Yuan-Mao) [reply cosim/black-parrot-example/v/top.v:490]: This change is from this separate PR https://github.com/black-parrot-hdk/zynq-parrot/pull/49. The first five commits of this PR are from other PRs so they should be removed when this PR gets merged.

File: cosim/black-parrot-example/v/top.v (hunk 11 above)
Comment (Yuan-Mao) [reply cosim/black-parrot-example/v/top.v:525]: This change is from this separate PR https://github.com/black-parrot-hdk/zynq-parrot/pull/49.

File: cosim/include/vcs/bp_zynq_pl.h (hunk 14 above)
Comment (Yuan-Mao) [reply cosim/include/vcs/bp_zynq_pl.h:56]: This change is from this separate PR https://github.com/black-parrot-hdk/zynq-parrot/pull/49.

File: cosim/black-parrot-example/ps.cpp (hunk 3 above)
Comment (Yuan-Mao) [reply cosim/black-parrot-example/ps.cpp:199]: It should be well defined. https://stackoverflow.com/questions/8391732/is-1u-valid-c

File: cosim/black-parrot-example/v/top_zynq.sv (hunk 13 above)
Comment (Yuan-Mao) [reply cosim/black-parrot-example/v/top_zynq.sv:213]: I agree, but then in this BP example we would have no components for which to generate resets. Should we move all the reset generation logic and the bit banging in ps.cpp to BP full example (with PLIC & Ethernet)?

File: cosim/black-parrot-example/ps.cpp (hunk 3 above)
Comment (dpetrisko) [reply cosim/black-parrot-example/ps.cpp:199]: cool!

File: cosim/black-parrot-example/v/top.v (hunk 10 above)
Comment (dpetrisko) [reply cosim/black-parrot-example/v/top.v:490]: yep, that's my bad. I think those PRs are merged now, can you do a rebase?

File: cosim/black-parrot-example/v/top_zynq.sv (hunk 13 above)
Comment (dpetrisko) [reply cosim/black-parrot-example/v/top_zynq.sv:213]: We can still have the BP reset, just not the AXI components

**Comment by Yuan-Mao** [comment]
This PR is closed as we have a new rebased version PR #51.
